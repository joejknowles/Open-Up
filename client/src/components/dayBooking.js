import React, { Component } from 'react';
import '../styles/App.css';
import { connect } from 'react-redux';

export class DayBooking extends Component {
  render() {
    const date = this.props.date ?
      this.props.date.toLocaleDateString("en-GB") :
      '';
    const slots = this.props.slots || []
    return (
      <div>
        <h3>Today { date }</h3>
        { slots.map(({ id }) => (
            <button key={ id } ></button>
        )) }
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  slots: state.allSlots.map((id) => state.slotsById[id]),
  date: state.selectedDate
});

export default connect(mapStateToProps)(DayBooking);
