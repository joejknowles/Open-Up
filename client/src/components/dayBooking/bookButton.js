import React from 'react';
import { connect } from 'react-redux';
import format from 'date-fns/format'
import { createSlotSelector } from '../../reducers'

const formatTime = (time) => (
  format(time, 'HH:mm')
);

export const BookButton = ({ startTime, endTime}) => (
  <button className="BookButton">book { formatTime(startTime) } to { formatTime(endTime) }</button>
);

const mapStateToProps = (state, { id }) => (
  createSlotSelector(id)
);

export default connect(mapStateToProps)(BookButton);
