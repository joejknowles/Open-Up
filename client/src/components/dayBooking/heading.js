import React from 'react';
import { connect } from 'react-redux';
import { selectedDateSelector } from '../../reducers';
import format from 'date-fns/format'
import { nextDay } from '../../actions/changeDay';

const dateTitle = (date) => (
  `Today ${format(date, 'D/M/YYYY')}`
);

export const Heading = ({ date, onNextDayClick }) => (
  <div className="Heading">
    <h3>{ dateTitle(date) }</h3>
    <a
      className="next-day"
      onClick={ onNextDayClick }>
      Next Day
    </a>
  </div>
);

export const mapStateToProps = (state) => ({
  date: selectedDateSelector(state)
});

const mapDispatchToProps = (dispatch) => ({ onNextDayClick: () => dispatch(nextDay) });

export default connect(mapStateToProps, mapDispatchToProps)(Heading);
