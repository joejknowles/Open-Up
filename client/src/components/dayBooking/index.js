import React, { Component } from 'react';
import '../../styles/App.css';
import { connect } from 'react-redux';
import { fetchSlots } from '../../actions';

import LoadingIndicator from '../loadingIndicator';
import Heading from './heading';
import List from './list';

export class DayBooking extends Component {
  componentDidMount() {
    this.props.fetchSlots();
  }

  render() {
    if (this.props.isLoading) return (<LoadingIndicator />);
    return (
      <div>
        <Heading />
        <List slots={ this.props.slots } />
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  slots: state.allSlots.map((id) => state.slotsById[id]),
  date: state.selectedDate,
  isLoading: state.isLoading
});

const mapDispatchToProps = (dispatch) => ({
  fetchSlots: fetchSlots(dispatch)
});

const mergeProps = ({ slots, isLoading, date }, dispatchProps) => ({
  slots, isLoading,
  fetchSlots: dispatchProps.fetchSlots(date)
});

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(DayBooking);
