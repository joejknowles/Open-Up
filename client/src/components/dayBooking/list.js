import React from 'react';
import '../../styles/DayBooking.css';
import { connect } from 'react-redux';
import { slotsSelector } from '../../reducers';
import Slot from './slot'

export const List = ({ slots = [] }) => (
  slots.length === 0 ?
  <p className="message">No slots available</p> :
  (<div className="List">
    { slots.map(( id ) => (
      <Slot key={ id } { ...{ id } } />
    )) }
  </div>)
);

const mapStateToProps = (state) => ({
  slots: slotsSelector(state)
});

const connectedList = connect(mapStateToProps)(List);

export default connectedList;
