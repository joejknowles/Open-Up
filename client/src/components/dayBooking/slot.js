import React from 'react';
import '../../styles/DayBooking.css';
import { connect } from 'react-redux';
import { createSlotSelector } from '../../reducers';
import BookButton from './bookButton';
import format from 'date-fns/format';

const formatTime = (time) => (
  format(time, 'HH:mm')
);

export const Slot = (props) => (
  <div className="Slot">
    {
      props.booking ?
        <UnavailableSlot { ...props } />
      :
        <BookButton { ...props }  />
    }
  </div>
);

export const UnavailableSlot = ({ startTime, endTime, booking }) => {
  const message = booking === "pending" ? '-' : `unavailable ${ formatTime(startTime) } to ${ formatTime(endTime) }`
  return (
    <div className="unavailableSlot slot-content">
      <span className="message">
        { message }
      </span>
    </div>
  );
}

const mapStateToProps = (state, { id }) => (
  createSlotSelector(id)
);

export default connect(mapStateToProps)(Slot);
