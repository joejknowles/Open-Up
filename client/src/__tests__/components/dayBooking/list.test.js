import { List } from '../../../components/dayBooking/list';

import React from 'react';
import { shallow } from 'enzyme';
import ReactTestUtils from 'react-addons-test-utils';

const shallowRenderer = ReactTestUtils.createRenderer();

it('renders two buttons for two slots', () => {
  const component = shallow(<List
    slots={ [{ id: 1 }, { id: 2 }] }
    fetchSlots={ ()=> {}}
    />);
  expect(component.find('button').length).toBe(2);
});

it('renders three buttons for three slots', () => {
  const component = shallow(<List
    slots={ [{ id: 1 }, { id: 2 }, { id: 4 }] }
    fetchSlots={ ()=> {}}
    />);
  expect(component.find('button').length).toBe(3);
});

it('renders message for no slots', () => {
  const component = shallow(<List
    slots={ [] }
    fetchSlots={ ()=> {}}
    />);
  expect(component.find('.message').text()).toBe('No slots available');
});

it('renders the same as before with two slots', () => {
  const tree = shallowRenderer.render(
    <List
      slots={ [{ id: 1 }, { id: 2 }] }
      fetchSlots={ ()=> {}}
      />
  );
  expect(tree).toMatchSnapshot();
});

it('renders the same as before with no slots', () => {
  const tree = shallowRenderer.render(
    <List
      slots={ [] }
      fetchSlots={ ()=> {}}
      />
  );
  expect(tree).toMatchSnapshot();
});
