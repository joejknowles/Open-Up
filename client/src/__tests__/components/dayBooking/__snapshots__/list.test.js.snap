exports[`test renders the same as before with no slots 1`] = `
<p
  className="message">
  No slots available
</p>
`;

exports[`test renders the same as before with two slots 1`] = `
<div
  className="List">
  <div
    className="Slot">
    <Connect(Component)
      id={1} />
  </div>
  <div
    className="Slot">
    <Connect(Component)
      id={2} />
  </div>
</div>
`;
